engine/table/src/main/java/io/deephaven/engine/table/impl/sources/UnionSourceManager.java

```diff
@@ -556,8 +558,13 @@ private void processExisting(@NotNull final Table constituent) {
                     resultRows.insertWithShift(currFirstRowKey, constituent.getRowSet());
                 }
             } else if (shiftDelta != 0) {
-                Assert.assertion(slotAllocationChanged, "slotAllocationChanged");
-                downstreamShiftBuilder.shiftRange(prevFirstRowKey, prevLastRowKey, shiftDelta);
+                try (final RowSet prevRowSet = constituent.getRowSet().copyPrev()) {
+                    // if prev row set is identical to rows removed, then we do not need to shift
+                    if (!prevRowSet.equals(changes.removed())) {
+                        Assert.assertion(slotAllocationChanged, "slotAllocationChanged");
+                        downstreamShiftBuilder.shiftRange(prevFirstRowKey, prevLastRowKey, shiftDelta);
+                    }
+                }
```

Comment:
This could be cheaper, and the cheaper version can be written more cleanly:
```
            } else if (shiftDelta != 0 && constituent.getRowSet().sizePrev() != changes.removed().size()) {
                // Note that changes.removed() must be a subset of the constituent's previous row set.
                // If constituent is removing all of its previous rows, then we do not need to propagate a shift.
                Assert.assertion(slotAllocationChanged, "slotAllocationChanged");
                downstreamShiftBuilder.shiftRange(prevFirstRowKey, prevLastRowKey, shiftDelta);
            }
```